File: src/sim/World.ts
/**
 * Headless game world: ALL simulation state as plain, serializable data, with no
 * Phaser/render dependency. `stepWorld` advances one fixed tick driven only by
 * (world, input) — the property that makes a match replayable from a tape and
 * verifiable by re-execution (the Teeworlds-on-CKB model).
 *
 * The sim deliberately MUTATES a single owned WorldState; leaf math (e.g.
 * stepProjectile) stays pure.
 */
import { TerrainMask, generateTerrainMask } from '../terrain/TerrainGenerator';
import { isSolid, carveCircle, columnSurface } from '../physics/DestructibleTerrain';
import { stepProjectile, ProjectileState, Vec2 } from '../physics/ProjectilePhysics';
import { WEAPONS, WEAPON_ORDER, weaponAt } from '../weapons/weaponData';
import {
  AimState, createAim, aimAngle, adjustElevation, setFacing, startCharge, updateCharge, release,
} from '../core/aim';
import { nextRandom } from '../core/rng';
import { dcos, dsin } from '../core/trig';
import { FIXED_DT } from '../core/time';
import { blake2b } from '@noble/hashes/blake2.js';
import { serializeWorld, CKB_HASH_PERSONAL } from './serialize';

export const APE_GRAVITY = 900; // px/s^2 — apes fall faster than projectiles
export const APE_WIDTH = 24;
export const APE_HEIGHT = 36;
export const MAX_WIND = 220; // px/s^2 at full strength
export const APE_WALK_SPEED = 80; // px/s horizontal walk speed (active ape, grounded)
export const APE_JUMP_SPEED = 420; // px/s upward launch on jump
export const MAX_STEP = 10; // max surface rise (px) the walking ape climbs per tick
export const WALK_BUDGET = 112.5; // px of movement allotted per turn (resets each turn)
export const JUMP_COST = 40; // px a jump draws from the turn's movement budget
const JUMP_LIFT = 2; // px nudge to clear the ground probe so a jump goes airborne
const SHOT_SUBSTEPS = 4; // anti-tunnelling sub-steps per tick

// --- P2 turn-loop constants ---
export const APES_PER_TEAM = 3;
export const APE_MAX_HEALTH = 100;
export const TURN_TICKS = 1500; // 30 s @ 50 Hz aiming budget
export const RESOLVE_MAX_TICKS = 400; // 8 s spiral guard for the settle wait
export const KNOCKBACK = 320; // px/s impulse at blast centre
export const FALL_DAMAGE_THRESHOLD = 600; // px/s landing speed before damage
export const FALL_DAMAGE_SCALE = 0.05; // health lost per px/s over the threshold
const GROUND_FRICTION = 0.7; // grounded horizontal velocity decay per tick
const REST_EPSILON = 1; // |velX| below this snaps to 0 (lets the world reach rest)

export type Phase = 'AIMING' | 'RESOLVING' | 'TURN_END' | 'GAMEOVER';

export interface ApeState {
  team: number;   // 0 or 1
  health: number; // starts APE_MAX_HEALTH; dead when <= 0
  x: number;
  y: number;
  prevX: number;  // previous tick position (render interpolation)
  prevY: number;
  velX: number;   // knockback / airborne horizontal velocity
  velY: number;
}

export interface ShotState {
  state: ProjectileState;
  prevPos: Vec2;
  weapon: number; // WEAPON_ORDER index this shot was fired with
}

export type SimEvent = { type: 'detonation'; x: number; y: number; radius: number };

export interface WorldState {
  width: number;
  height: number;
  tick: number;
  rng: number;
  wind: number;
  aim: AimState;
  apes: ApeState[];
  activeApe: number;
  phase: Phase;
  turnTimer: number;     // ticks left in AIMING
  resolveTimer: number;  // ticks elapsed in RESOLVING (spiral guard)
  moveBudget: number;    // px of movement left this turn (walk drains it, jump costs a chunk)
  teamNext: [number, number]; // next roster position to act, per team
  winner: number | null; // team index, -1 draw, null ongoing
  selectedWeapon: number; // sticky WEAPON_ORDER index, default 0
  ammo: number[][];       // ammo[team][weaponIndex]; -1 = unlimited
  shot: ShotState | null;
  mask: TerrainMask;
  events: SimEvent[];
}

export interface TickInput {
  aimUp: boolean;
  aimDown: boolean;
  aimLeft?: boolean;  // face left this tick
  aimRight?: boolean; // face right this tick
  moveLeft?: boolean;   // walk left this tick (active ape, grounded, AIMING)
  moveRight?: boolean;  // walk right this tick
  jumpPressed?: boolean; // jump this tick (edge-triggered; grounded only)
  fireHeld: boolean;
  firePressed: boolean;
  fireReleased: boolean;
  selectWeapon?: number; // set only on the tick a selection is confirmed
}

/** True if an ape is still in play. */
export function alive(ape: ApeState, height: number): boolean {
  return ape.health > 0 && ape.y <= height;
}

/** Global ape indices belonging to a team, in placement order. */
export function teamApeIndices(world: WorldState, team: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < world.apes.length; i++) if (world.apes[i].team === team) out.push(i);
  return out;
}

export function createWorld(seed: number, width: number, height: number): WorldState {
  const mask = generateTerrainMask(width, height, seed);
  // Teams spawn on OPPOSING sides: team 0 clustered on the left, team 1 mirrored
  // on the right. Index order is contiguous per team (team 0 first), so
  // activeApe 0 is team 0's lead and teamApeIndices stays in placement order.
  const SPAWN_MARGIN = 0.10; // nearest ape to the edge
  const SPAWN_SPAN = 0.28;   // how far the team spreads inward from the margin
  const apes: ApeState[] = [];
  for (let team = 0; team < 2; team++) {
    for (let j = 0; j < APES_PER_TEAM; j++) {
      const t = j / Math.max(1, APES_PER_TEAM - 1); // 0..1 across the team (guards 1-ape teams)
      const fromEdge = SPAWN_MARGIN + SPAWN_SPAN * t; // 0.10 .. 0.38 from the team's edge
      const frac = team === 0 ? fromEdge : 1 - fromEdge;
      const x = Math.floor(width * frac);
      const surfaceY = columnSurface(mask, x) ?? height - 50;
      const y = surfaceY - APE_HEIGHT / 2;
      apes.push({ team, health: APE_MAX_HEALTH, x, y, prevX: x, prevY: y, velX: 0, velY: 0 });
    }
  }

  const roll = nextRandom(seed >>> 0);
  const startAmmo = WEAPON_ORDER.map((id) => WEAPONS[id].ammoStart);
  return {
    width,
    height,
    tick: 0,
    rng: roll.next,
    wind: (roll.value * 2 - 1) * MAX_WIND,
    aim: createAim(1),        // team 0 starts on the left, facing right toward the enemy
    apes,
    activeApe: 0,             // team 0's first ape
    phase: 'AIMING',
    turnTimer: TURN_TICKS,
    resolveTimer: 0,
    moveBudget: WALK_BUDGET,
    teamNext: [1, 0],        // team 0 already acting at pos 0; next is pos 1
    winner: null,
    selectedWeapon: 0,
    ammo: [startAmmo.slice(), startAmmo.slice()],
    shot: null,
    mask,
    events: [],
  };
}

/** Logical muzzle of the active ape. */
export function muzzle(world: WorldState): Vec2 {
  const ape = world.apes[world.activeApe];
  const angle = aimAngle(world.aim);
  const clearance = 22;
  return {
    x: ape.x + dcos(angle) * clearance,
    y: ape.y - APE_HEIGHT / 2 - dsin(angle) * clearance,
  };
}

/** Advance the world exactly one fixed tick. (Behavior fleshed out in Tasks 2–5.) */
export function stepWorld(world: WorldState, input: TickInput): void {
  world.events.length = 0;

  if (world.phase === 'GAMEOVER') {
    settleApes(world, 0);
    world.tick++;
    return;
  }

  // Direction the active ape is walking this tick (0 = not walking). Threaded into
  // settleApes so only deliberate walking step-climbs slopes; knockback never does.
  let walkDir = 0;

  if (world.phase === 'AIMING') {
    if (input.selectWeapon !== undefined) {
      const i = input.selectWeapon;
      const team = world.apes[world.activeApe].team;
      if (i >= 0 && i < WEAPON_ORDER.length && world.ammo[team][i] !== 0) {
        world.selectedWeapon = i;
      }
    }
    const aim = world.aim;

    // Active-ape locomotion (its turn only). Walking drives velX directly; ground
    // friction in settleApes only slows the ape once input stops (a short slide).
    // Jumping nudges y up off the ground probe so settleApes goes airborne WITHOUT
    // changing the shared settle physics — kept byte-identical to the on-chain verifier.
    const mover = world.apes[world.activeApe];
    if (isSolid(world.mask, mover.x, mover.y + APE_HEIGHT / 2 + 1)) {
      const dir = (input.moveRight ? 1 : 0) - (input.moveLeft ? 1 : 0);
      // Jump first so the walk step below can't nibble the budget a jump needs.
      if (input.jumpPressed && world.moveBudget >= JUMP_COST) {
        mover.velY = -APE_JUMP_SPEED;
        mover.y -= JUMP_LIFT;
        world.moveBudget -= JUMP_COST;
      }
      if (dir !== 0 && world.moveBudget > 0) {
        mover.velX = dir * APE_WALK_SPEED;
        setFacing(aim, dir); // face the way we walk
        walkDir = dir;
        world.moveBudget -= APE_WALK_SPEED * FIXED_DT; // drain the allotted step
      }
    }

    if (input.aimUp) adjustElevation(aim, 1, FIXED_DT);
    if (input.aimDown) adjustElevation(aim, -1, FIXED_DT);
    if (input.aimLeft) setFacing(aim, -1);
    if (input.aimRight) setFacing(aim, 1);
    if (!world.shot) {
      if (input.firePressed) startCharge(aim);
      if (input.fireHeld) updateCharge(aim, FIXED_DT);
      if (input.fireReleased) fire(world, release(aim));
      if (world.shot) {
        // Shot just launched — enter RESOLVING immediately.
        world.phase = 'RESOLVING';
        world.resolveTimer = 0;
      } else {
        world.turnTimer--;
        if (world.turnTimer <= 0) world.phase = 'TURN_END';
      }
    }
  }

  advanceShot(world);
  settleApes(world, walkDir);

  if (world.phase === 'RESOLVING') {
    world.resolveTimer++;
    if (worldAtRest(world) || world.resolveTimer >= RESOLVE_MAX_TICKS) {
      world.phase = 'TURN_END';
    }
  }

  if (world.phase === 'TURN_END') endTurn(world);

  world.tick++;
}

/** No shot in flight and every living ape is motionless. */
function worldAtRest(world: WorldState): boolean {
  if (world.shot) return false;
  for (const ape of world.apes) {
    if (!alive(ape, world.height)) continue;
    if (ape.velX !== 0 || ape.velY !== 0) return false;
  }
  return true;
}

function countAlive(world: WorldState, team: number): number {
  let n = 0;
  for (const ape of world.apes) if (ape.team === team && alive(ape, world.height)) n++;
  return n;
}

/** Rotate to the next ape on the other team and start a fresh AIMING turn. */
function endTurn(world: WorldState): void {
  world.shot = null; // discard any projectile still in flight if the resolve guard fired
  const a0 = countAlive(world, 0);
  const a1 = countAlive(world, 1);
  if (a0 === 0 || a1 === 0) {
    world.winner = a0 === 0 && a1 === 0 ? -1 : a0 === 0 ? 1 : 0;
    world.phase = 'GAMEOVER';
    return;
  }
  const nextTeam = 1 - world.apes[world.activeApe].team;
  world.activeApe = nextLivingApeOnTeam(world, nextTeam);
  rerollTurn(world);
  world.phase = 'AIMING';
}

/** Next LIVING ape on a team, advancing the per-team cursor and skipping corpses. */
function nextLivingApeOnTeam(world: WorldState, team: number): number {
  const roster = teamApeIndices(world, team);
  const start = world.teamNext[team] % roster.length;
  for (let k = 0; k < roster.length; k++) {
    const pos = (start + k) % roster.length;
    const idx = roster[pos];
    if (alive(world.apes[idx], world.height)) {
      world.teamNext[team] = (pos + 1) % roster.length;
      return idx;
    }
  }
  return world.activeApe; // unreachable: win check guarantees a living ape exists
}

function rerollTurn(world: WorldState): void {
  const roll = nextRandom(world.rng);
  world.rng = roll.next;
  world.wind = (roll.value * 2 - 1) * MAX_WIND;
  // Default facing toward the enemy: team 0 (left) faces right, team 1 (right) faces left.
  world.aim = createAim(world.apes[world.activeApe].team === 0 ? 1 : -1);
  world.turnTimer = TURN_TICKS;
  world.resolveTimer = 0;
  world.moveBudget = WALK_BUDGET;
}

function fire(world: WorldState, power: number): void {
  if (power <= 0) return;
  const i = world.selectedWeapon;
  const team = world.apes[world.activeApe].team;
  if (world.ammo[team][i] === 0) return; // empty: cannot fire
  const weapon = weaponAt(i);
  const speed = power * weapon.launchSpeed;
  const angle = aimAngle(world.aim);
  const m = muzzle(world);
  world.shot = {
    prevPos: { x: m.x, y: m.y },
    state: {
      pos: { x: m.x, y: m.y },
      vel: { x: dcos(angle) * speed, y: -dsin(angle) * speed },
    },
    weapon: i,
  };
  if (world.ammo[team][i] > 0) {
    world.ammo[team][i]--;
    // When the last round of a finite weapon is spent, revert the sticky selection
    // to moonShot (index 0) so the next turn's fire trigger is never dead.
    if (world.ammo[team][i] === 0) world.selectedWeapon = 0;
  }
}

function settleApes(world: WorldState, walkDir: number): void {
  for (let i = 0; i < world.apes.length; i++) {
    const ape = world.apes[i];
    ape.prevX = ape.x;
    ape.prevY = ape.y;
    if (!alive(ape, world.height)) continue; // dead apes don't move

    // Only deliberate walking step-climbs; knockback/idle keep the plain wall probe
    // so blast physics stays byte-identical to before (and to the on-chain verifier).
    const walking = i === world.activeApe && walkDir !== 0;

    // Horizontal: integrate velX.
    if (ape.velX !== 0) {
      const nx = ape.x + ape.velX * FIXED_DT;
      if (walking && isSolid(world.mask, ape.x, ape.y + APE_HEIGHT / 2 + 1)) {
        // Grounded walk: snap the feet onto the surface at nx so the ape climbs
        // smooth slopes instead of burrowing into them (the mid-height probe alone
        // only stops it once the ground rises ~19px). A rise over MAX_STEP is a wall.
        const surf = columnSurface(world.mask, nx);
        if (surf === null) {
          if (!isSolid(world.mask, nx, ape.y)) ape.x = nx; // step off into a gap
          else ape.velX = 0;
        } else {
          const targetY = surf - APE_HEIGHT / 2 - 1; // resting y: feet 1px above surf
          const climb = ape.y - targetY;             // >0 = ground rose (uphill)
          if (climb > MAX_STEP) ape.velX = 0;        // too steep — a wall
          else if (climb > 0) { ape.x = nx; ape.y = targetY; } // step up onto it
          else if (!isSolid(world.mask, nx, ape.y)) ape.x = nx; // flat / downhill
          else ape.velX = 0;
        }
      } else if (isSolid(world.mask, nx, ape.y)) {
        ape.velX = 0; // stop dead at a solid wall (probe at mid-height)
      } else {
        ape.x = nx;
      }
    }

    // Vertical: gravity while airborne; on landing apply fall damage + friction.
    const feetY = ape.y + APE_HEIGHT / 2;
    if (!isSolid(world.mask, ape.x, feetY + 1)) {
      ape.velY += APE_GRAVITY * FIXED_DT;
      ape.y += ape.velY * FIXED_DT;
    } else {
      if (ape.velY > FALL_DAMAGE_THRESHOLD) {
        ape.health -= FALL_DAMAGE_SCALE * (ape.velY - FALL_DAMAGE_THRESHOLD);
      }
      ape.velY = 0;
      ape.velX *= GROUND_FRICTION;
      if (Math.abs(ape.velX) < REST_EPSILON) ape.velX = 0;
    }

    // Water: clamp a fallen ape to a stable sentinel y (it's dead via alive()).
    if (ape.y > world.height + 50) {
      ape.y = world.height + 50;
      ape.velX = 0;
      ape.velY = 0;
    }
  }
}

function advanceShot(world: WorldState): void {
  if (!world.shot) return;
  world.shot.prevPos = { x: world.shot.state.pos.x, y: world.shot.state.pos.y };
  const weapon = weaponAt(world.shot.weapon);
  const sub = FIXED_DT / SHOT_SUBSTEPS;
  for (let i = 0; i < SHOT_SUBSTEPS; i++) {
    world.shot.state = stepProjectile(world.shot.state, weapon.projectile, world.wind, sub);
    const { x, y } = world.shot.state.pos;
    const offscreen = x < -50 || x > world.width + 50 || y > world.height + 50;
    if (isSolid(world.mask, x, y) || offscreen) {
      if (!offscreen) detonate(world, x, y, weapon.blastRadius);
      world.shot = null;
      return;
    }
  }
}

function detonate(world: WorldState, x: number, y: number, radius: number): void {
  carveCircle(world.mask, x, y, radius);
  world.events.push({ type: 'detonation', x, y, radius });
  const weapon = world.shot ? weaponAt(world.shot.weapon) : weaponAt(0);
  applyBlast(world, x, y, radius, weapon.damage);
}

/** Radial falloff damage + knockback to every living ape within `radius`. */
function applyBlast(world: WorldState, x: number, y: number, radius: number, damage: number): void {
  for (const ape of world.apes) {
    if (!alive(ape, world.height)) continue;
    const dx = ape.x - x;
    const dy = ape.y - y;
    const d = Math.sqrt(dx * dx + dy * dy);
    if (d > radius) continue;
    const falloff = 1 - d / radius;
    ape.health -= damage * falloff;
    let nx: number;
    let ny: number;
    if (d === 0) {
      nx = 0;
      ny = -1; // dead-centre: launch straight up
    } else {
      nx = dx / d;
      ny = dy / d;
    }
    const impulse = KNOCKBACK * falloff;
    ape.velX += nx * impulse;
    ape.velY += ny * impulse;
  }
}

/** Test seam: drive a blast directly without scripting a full shot. */
export function detonateAt(world: WorldState, x: number, y: number, radius: number, damage: number): void {
  carveCircle(world.mask, x, y, radius);
  applyBlast(world, x, y, radius, damage);
}

/**
 * 32-byte cryptographic commitment to the full world: blake2b-256 over the
 * canonical serialization, using CKB's ckbhash personalization so an on-chain
 * CKB-VM verifier reproduces the exact digest via the chain's native blake2b.
 * Verification is re-execution: replay the tape, commit, compare to the claim.
 */
export function commitWorld(world: WorldState): Uint8Array {
  return blake2b(serializeWorld(world), { dkLen: 32, personalization: CKB_HASH_PERSONAL });
}

File: verifier/src/world.rs
//! Rust mirror of the TypeScript `WorldState` and a `serialize_world` that is
//! BYTE-IDENTICAL to `src/sim/serialize.ts`'s `serializeWorld`.
//!
//! Field order here is LOAD-BEARING and mirrors `serialize.ts` line-for-line —
//! it must never change without invalidating every past tape and breaking the
//! on-chain digest. Floats go through [`crate::quantize`] (JS `Math.round`
//! semantics) and u32s use two's-complement reinterpret (JS `n >>> 0`) so that
//! `-1` encodes as `0xFFFFFFFF` and `winner == null` encodes as `99`.

use crate::aim::{create_aim, AimState};
use crate::physics::ProjectileState;
use crate::terrain::generate_terrain_mask;
use crate::weapons::{weapon_at, WEAPON_COUNT};
use crate::{column_surface, quantize};
use serde::Deserialize;
use std::fs;

// --- createWorld constants, ported from src/sim/World.ts ---
const APES_PER_TEAM: usize = 3;
const APE_MAX_HEALTH: f64 = 100.0;
const APE_HEIGHT: f64 = 36.0;
const MAX_WIND: f64 = 220.0;
const TURN_TICKS: i64 = 1500;
const SPAWN_MARGIN: f64 = 0.10;
const SPAWN_SPAN: f64 = 0.28;

/// Phase strings in the canonical order; the index is what gets serialized.
/// Mirrors `PHASE_ORDER` / `indexOf` in `serialize.ts` (missing -> -1).
const PHASE_ORDER: [&str; 4] = ["AIMING", "RESOLVING", "TURN_END", "GAMEOVER"];

fn phase_index(phase: &str) -> i64 {
    PHASE_ORDER
        .iter()
        .position(|p| *p == phase)
        .map(|i| i as i64)
        .unwrap_or(-1)
}

/// One ape. Mirrors the serialized subset of `ApeState` in `src/sim/World.ts`
/// (`prevX`/`prevY` exist in the JSON but are not serialized, so they are
/// dropped — serde ignores unknown fields by default).
#[derive(Debug, Deserialize)]
pub struct ApeState {
    pub team: i64,
    pub health: f64,
    pub x: f64,
    pub y: f64,
    #[serde(rename = "velX")]
    pub vel_x: f64,
    #[serde(rename = "velY")]
    pub vel_y: f64,
}

/// In-flight shot. Mirrors the serialized subset of `ShotState` (`prevPos` is
/// in the JSON but not serialized, so it is ignored).
#[derive(Debug, Deserialize)]
pub struct ShotState {
    pub state: ProjectileState,
    pub weapon: i64,
}

/// Mirror of the TS `WorldState`, holding only the fields `serialize_world`
/// reads (plus the raw terrain `mask`, which is loaded separately from a binary
/// sidecar rather than from JSON). Unknown JSON fields (`width`, `height`,
/// `prevX`/`prevY`, `events`, ...) are silently ignored by serde.
#[derive(Debug, Deserialize)]
pub struct WorldState {
    pub tick: i64,
    pub rng: i64,
    pub phase: String,
    #[serde(rename = "activeApe")]
    pub active_ape: i64,
    #[serde(rename = "turnTimer")]
    pub turn_timer: i64,
    #[serde(rename = "resolveTimer")]
    pub resolve_timer: i64,
    /// `null` while the match is ongoing; `-1` for a draw; otherwise a team
    /// index. Encoded as `winner ?? 99`.
    pub winner: Option<i64>,
    #[serde(rename = "teamNext")]
    pub team_next: [i64; 2],
    pub wind: f64,
    pub apes: Vec<ApeState>,
    /// Unified game-logic + serde aim type (see [`crate::aim::AimState`]).
    pub aim: AimState,
    #[serde(rename = "selectedWeapon")]
    pub selected_weapon: i64,
    /// `ammo[team][weaponIndex]`; `-1` means unlimited (u32-encoded as
    /// `0xFFFFFFFF`).
    pub ammo: Vec<Vec<i64>>,
    pub shot: Option<ShotState>,
    /// Raw terrain mask bytes, appended verbatim. Populated by
    /// [`load_fixture_world`], not deserialized from JSON.
    #[serde(skip)]
    pub mask: Vec<u8>,
}

/// Growable little-endian byte writer mirroring `ByteWriter` in `serialize.ts`.
struct ByteWriter {
    buf: Vec<u8>,
}

impl ByteWriter {
    fn new() -> Self {
        Self { buf: Vec::new() }
    }

    /// Unsigned 32-bit LE via two's-complement reinterpret. `n as u32` keeps the
    /// low 32 bits, exactly matching JS `n >>> 0` (`-1` -> `0xFFFFFFFF`).
    fn u32(&mut self, n: i64) {
        self.buf.extend_from_slice(&(n as u32).to_le_bytes());
    }

    /// Canonical fixed-point float: `quantize(v)` (JS `Math.round(v*1000)`) as a
    /// signed 64-bit LE value. Uses [`crate::quantize`], NOT `f64::round`.
    fn f(&mut self, v: f64) {
        self.buf.extend_from_slice(&quantize(v).to_le_bytes());
    }

    /// Raw bytes, verbatim (the terrain mask).
    fn bytes(&mut self, b: &[u8]) {
        self.buf.extend_from_slice(b);
    }
}

/// Serialize a [`WorldState`] to its canonical byte form — byte-identical to the
/// TypeScript `serializeWorld`. Field order mirrors `serialize.ts` exactly.
pub fn serialize_world(world: &WorldState) -> Vec<u8> {
    let mut w = ByteWriter::new();

    w.u32(world.tick);
    w.u32(world.rng);
    w.u32(phase_index(&world.phase));
    w.u32(world.active_ape);
    w.u32(world.turn_timer);
    w.u32(world.resolve_timer);
    w.u32(world.winner.unwrap_or(99));
    w.u32(world.team_next[0]);
    w.u32(world.team_next[1]);
    w.f(world.wind);

    for ape in &world.apes {
        w.u32(ape.team);
        w.f(ape.health);
        w.f(ape.x);
        w.f(ape.y);
        w.f(ape.vel_x);
        w.f(ape.vel_y);
    }

    w.u32(world.aim.facing as i64);
    w.f(world.aim.elevation);
    w.f(world.aim.power);
    w.u32(if world.aim.is_charging { 1 } else { 0 });
    w.u32(world.selected_weapon);

    for team in &world.ammo {
        for ammo in team {
            w.u32(*ammo);
        }
    }

    w.u32(if world.shot.is_some() { 1 } else { 0 });
    if let Some(shot) = &world.shot {
        w.f(shot.state.pos.x);
        w.f(shot.state.pos.y);
        w.f(shot.state.vel.x);
        w.f(shot.state.vel.y);
        w.u32(shot.weapon);
    }

    w.bytes(&world.mask);
    w.buf
}

/// Build the initial world natively from a seed — a byte-exact port of
/// `createWorld` in `src/sim/World.ts`. Generates the terrain, spawns both teams,
/// rolls the opening wind, and seeds aim/ammo. Proves full initial-state parity:
/// `serialize_world(create_world(...))` reproduces the TS commitment from the
/// seed alone (no fixture JSON).
pub fn create_world(seed: i32, width: i32, height: i32) -> WorldState {
    let mask = generate_terrain_mask(width, height, seed);

    // Teams spawn on opposing sides: team 0 left, team 1 mirrored right.
    // Index order is contiguous per team (team 0 first).
    let mut apes: Vec<ApeState> = Vec::with_capacity(2 * APES_PER_TEAM);
    for team in 0..2i64 {
        for j in 0..APES_PER_TEAM {
            // TS: t = j / Math.max(1, APES_PER_TEAM - 1)
            let t = j as f64 / ((APES_PER_TEAM as i64 - 1).max(1)) as f64;
            let from_edge = SPAWN_MARGIN + SPAWN_SPAN * t;
            let frac = if team == 0 {
                from_edge
            } else {
                1.0 - from_edge
            };
            let x = (width as f64 * frac).floor();
            let surface_y = column_surface(&mask, x).unwrap_or(height - 50);
            let y = surface_y as f64 - APE_HEIGHT / 2.0;
            apes.push(ApeState {
                team,
                health: APE_MAX_HEALTH,
                x,
                y,
                vel_x: 0.0,
                vel_y: 0.0,
            });
        }
    }

    // Wind roll. TS: nextRandom(seed >>> 0). `seed >>> 0` reinterprets the i32
    // seed bits as u32; next_random takes i32 and wraps identically, so passing
    // `seed` straight through is bit-identical to the TS cursor.
    let (value, next) = crate::next_random(seed);
    let start_ammo: Vec<i64> = (0..WEAPON_COUNT)
        .map(|i| weapon_at(i).ammo_start as i64)
        .collect();

    WorldState {
        tick: 0,
        rng: next as i64,
        phase: "AIMING".to_string(),
        active_ape: 0,
        turn_timer: TURN_TICKS,
        resolve_timer: 0,
        winner: None,
        team_next: [1, 0],
        wind: (value * 2.0 - 1.0) * MAX_WIND,
        apes,
        aim: create_aim(1), // team 0 starts on the left, facing right
        selected_weapon: 0,
        ammo: vec![start_ammo.clone(), start_ammo],
        shot: None,
        mask: mask.data,
    }
}

/// Load a fixture world: deserialize the struct (minus mask) from
/// `json_path`, then attach the raw mask bytes read from `mask_path`.
pub fn load_fixture_world(json_path: &str, mask_path: &str) -> WorldState {
    let json = fs::read_to_string(json_path).unwrap_or_else(|e| panic!("read {json_path}: {e}"));
    let mut world: WorldState =
        serde_json::from_str(&json).unwrap_or_else(|e| panic!("parse {json_path}: {e}"));
    world.mask = fs::read(mask_path).unwrap_or_else(|e| panic!("read {mask_path}: {e}"));
    world
}
